File: Исходники/Глава 3. Часть 1/1. Бинарный поиск/Доп. материалы для лекций/1. Бинарный поиск/C++/Продуктовый/binary search.cpp
template<typename T>
bool comp(T firstElement, T secondElement)
{
	return firstElement < secondElement;
}

template<typename T>
T binary_search(T arr[], int length, T data, bool comp)
{
	int left = 0;
	int right = length - 1;

	while (left <= right)
	{
		int mid = floor((left + right) / 2);
		
		if ( arr[mid] == data ) {return mid;}
		
		else if (comp(arr[mid], data)) {left = mid + 1;}
		
		else {right = mid - 1;}
	}
	
	return -1;
}

File: Исходники/Глава 3. Часть 1/1. Бинарный поиск/Доп. материалы для лекций/1. Бинарный поиск/Python/binary search.py
def binarySearch(array: list, data: int):
	left = 0
	right = len(array) - 1
	
	while left <= right:
		mid = (left + right) // 2
	
		if array[mid] == data:
			return mid
	
		elif array[mid] < data:
			left = mid + 1
	
		elif array[mid] > data:
			right = mid - 1
	
	return None